File: backend-java-api/src/main/java/com/example/backend_java_api/models/Gallery.java
package com.example.backend_java_api.models;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

// Represents a Gallery
public class Gallery {

    private String name;
    private List<Artwork> artworks;
    private int displayWallWeightCapacity; // Maximum weight that can be displayed on the wall (in grams)
    private int displayFloorWeightCapacity; // Maximum weight that can be displayed on the floor (in grams)
    private final int maxCarryWeight = 10000; // Maximum weight that can be carried by a person (in grams) - 10 kg

    // Constructor
    public Gallery(
        String name,
        int displayWallWeightCapacity,
        int displayFloorWeightCapacity
    ) {
        this.name = name;
        this.artworks = new ArrayList<>();
        this.displayWallWeightCapacity = displayWallWeightCapacity;
        this.displayFloorWeightCapacity = displayFloorWeightCapacity;
    }

    // Add a new artwork to the gallery
    public void addArtwork(String title, String artist, int year, int weight) {
        if (artworks.stream().anyMatch(a -> a.getTitle().equals(title))) {
            throw new IllegalArgumentException(
                "Artwork with title " +
                title +
                " already exists in the gallery."
            );
        }
        Artwork artwork = new Artwork(title, artist, year, weight);
        artworks.add(artwork);
    }

    // Add a list of artworks to the gallery
    public void addListOfArtworks(List<Artwork> newArtworks) {
        for (Artwork artwork : newArtworks) {
            addArtwork(
                artwork.getTitle(),
                artwork.getArtist(),
                artwork.getYear(),
                artwork.getWeight()
            );
        }
    }

    /**
     * Moves an artwork to either display or storage.
     *
     * @param artworkTitle The title of the artwork.
     * @param area The area to move the artwork to. Either 'display' or 'storage'.
     * @return true if the artwork was moved successfully, false otherwise.
     * @throws IllegalArgumentException if the area is not 'display' or 'storage'.
     * @description Moves an artwork to the specified area.
     * If moved to display, ensure that the new weight does not exceed the display area's capacity.
     */
    public boolean moveArtwork(String artworkTitle, String area) {
        // TODO: Implement this method
        return false;
    }

    // Getters and Setters--------------------------------------------
    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public List<Artwork> getArtworks() {
        return artworks;
    }

    public void setArtworks(List<Artwork> artworks) {
        this.artworks = artworks;
    }

    public int getDisplayWallWeightCapacity() {
        return displayWallWeightCapacity;
    }

    // IGNORE COMPLETELY: Find the minimum number of trips required to move all artworks
    public List<Integer> minimumNumberOfTrips() {
        return new ArrayList<>(); // Dummy implementation
    }
}

File: backend-node-api/models/gallery.js
import { Artwork } from "./artwork.js";

/**
 * Represents a gallery.
 * @class
 */
export class Gallery {
  /**
   * Creates a new gallery.
   */
  constructor(name, displayWallWeightCapacity, displayFloorWeightCapacity) {
    this.name = name;
    this.artworks = [];
    this.displayWallWeightCapacity = displayWallWeightCapacity; // Maximum weight that can be displayed on the wall in grams
    this.displayFloorWeightCapacity = displayFloorWeightCapacity; // Maximum weight that can be displayed on the floor in grams
    this.maxCarryWeight = 10000; // Maximum weight that can be carried by a person in grams - 10 kg
  }

  /**
   * @function addArtwork
   * @param {string} title - The title of the painting.
   * @param {string} artist - The artist who created the painting.
   * @param {number} year - The year the painting was created.
   * @param {number} weight - The weight of the painting in grams.
   * @description Function to add a new artwork to the gallery.
   */
  addArtwork(title, artist, year, weight) {
    if (this.artworks.some((artwork) => artwork.title === title)) {
      console.log(`Artwork with title ${title} already exists in the gallery.`);
      throw new Error(
        `Artwork with title ${title} already exists in the gallery.`,
      );
    }
    let artwork = new Artwork(title, artist, year, weight);
    this.artworks.push(artwork);
  }

  /**
    * @function addListOfArtworks
    * @param {Array} artworks - An array of artworks to be added to the gallery.
    * @description Function to add a list of artworks to the gallery.
    a list of artworks to the gallery.
  */
  addListOfArtworks(artworks) {
    artworks.forEach((artwork) => {
      this.addArtwork(
        artwork.title,
        artwork.artist,
        artwork.year,
        artwork.weight,
      );
    });
  }

  /**
   * Move a artwork to display or storage.
   * @function moveArtwork
   * @param {string} artworkTitle - The supposed title of the artwork.
   * @param {string} area - The area to move the artwork to. Either 'display' or 'storage'.
   * @returns {boolean} - True if the artwork was moved successfully, false otherwise.
   * @description Move a artwork to display or storage.
   * If the artwork is moved to display, the new weight of the display area should not exceed the display weight capacity.
   */
  moveArtwork(artworkTitle, area) {
    // TODO: Implement this method
    return false;
  }

  /**
   * @function minimumNumberOfTrips
   * @description Function to find the minimum number of trips required to move all the artworks to the gallery.
   * @returns {number[]} - The weight of carried artworks in each trip.
   * @description The total weight of the carried artworks in each trip should not exceed the maximum carry weight. The goal is to move all the artworks to the gallery in minimum number of trips.
   */
  minimumNumberOfTrips() {
    return [];
  }
}
